src/arch/riscv/pagetable_walker.cc

```diff
@@ -195,6 +223,41 @@ Walker::WalkerState::initState(ThreadContext * _tc,
     assert(satp.mode == AddrXlateMode::SV39);
 }
 
+std::pair<bool, Fault>
+Walker::WalkerState::tryCoalesce(ThreadContext *_tc,
+    BaseMMU::Translation *translation,
+    const RequestPtr &_req, BaseMMU::Mode _mode)
+{
+    SATP _satp = _tc->readMiscReg(MISCREG_SATP);
+    bool priv_match = (
+        (mode == _mode) && (satp == _satp) &&
+        (pmode == walker->tlb->getMemPriv(_tc, _mode)) &&
+        (status == _tc->readMiscReg(MISCREG_STATUS)));
+
+    bool addr_match = false;
+    Addr addr_match_num = req->getVaddr();
+    Addr pre_match_num = _req->getVaddr();
+
+    addr_match = ((pre_match_num >> PageShift) << PageShift) ==
+                 ((addr_match_num >> PageShift) << PageShift);
+
+    if (priv_match && addr_match) {
+        // coalesce
+        DPRINTF(PageTableWalker,
+        "Coalescing walk for %#lx(pc=%#lx) into %#lx(pc=%#lx)\n",
+        req->getVaddr(), _req->getPC(), req->getVaddr(), req->getPC());
+        // add to list of requestors
```

Comment:
Where is the list of requestors?

Is there a maximum size for this list? In real hardware, there would be a maximum number of transactions that could be coalesced. Are we modeling this? If so, how?

Reply:
The list of requestors is `currStates`.

In hypervisor mode (RISC-V H Extension), the TLB request will be more frequent, while multiple in-flight TLB requests could co-exist. This should be orthogonal work to squashing already hit TLB https://github.com/gem5/gem5/pull/672.

This pr aims to improve the in-flight TLB request performance, as mentioned above wait for H Extension pr, expect to work like Miss Status Handling Registers (MSHR) for TLB (Still not yet finished).

Yes, the maximum size should be considered, but not yet been decided.

Reply:
@powerjg I think the real problem here is that the PTW can hold infinite number of requests.

For coalescing, we can do this at either when we enqueue a request, or when a walk returns, and we can definitely limit the number of requests that can return (to the core, etc.) per cycle.